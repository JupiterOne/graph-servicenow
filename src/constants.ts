import { RelationshipClass } from '@jupiterone/integration-sdk-core';

export const Steps = {
  ACCOUNT: 'step-account',
  USERS: 'step-users',
  GROUPS: 'step-groups',
  GROUP_MEMBERS: 'step-group-members',
  INCIDENTS: 'step-incidents',
  CMDB: 'step-cmdb',
  USER_OWNS_CMDB: 'step_user_owns_cmdb',
  USER_MANAGES_CMDB: 'step_user_manages_cmdb',
  GROUP_MANAGES_CMDB: 'step_group_manages_cmdb',
  CMDB_ASSIGNED_USER: 'step_cmdb_assigned_user',
};

export const Entities = {
  ACCOUNT: {
    resourceName: 'Account',
    _type: 'service_now_account',
    _class: 'Account',
  },
  USER: {
    resourceName: 'User',
    _type: 'service_now_user',
    _class: 'User',
  },
  GROUP: {
    resourceName: 'User Group',
    _type: 'service_now_group',
    _class: 'UserGroup',
  },
  INCIDENT: {
    resourceName: 'Incident',
    _type: 'service_now_incident',
    _class: 'Incident',
  },
  CMDB_OBJECT: {
    resourceName: 'CMDB Object',
    _type: 'service_now_cmdb_object',
    _class: 'Configuration',
    disableClassMatch: true,
  },
};

export const Relationships = {
  ACCOUNT_HAS_USER: {
    _type: 'service_now_account_has_user',
    sourceType: Entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.USER._type,
  },
  ACCOUNT_HAS_GROUP: {
    _type: 'service_now_account_has_group',
    sourceType: Entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.GROUP._type,
  },
  GROUP_HAS_USER: {
    _type: 'service_now_group_has_user',
    sourceType: Entities.GROUP._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.USER._type,
  },
  GROUP_HAS_GROUP: {
    _type: 'service_now_group_has_group',
    sourceType: Entities.GROUP._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.GROUP._type,
  },
  INCIDENT_ASSIGNED_USER: {
    _type: 'service_now_incident_assigned_user',
    sourceType: Entities.INCIDENT._type,
    _class: RelationshipClass.ASSIGNED,
    targetType: Entities.USER._type,
  },
  CMDB_ASSIGNED_TO_USER: {
    _type: 'service_now_cmdb_object_assigned_user',
    sourceType: Entities.CMDB_OBJECT._type,
    _class: RelationshipClass.ASSIGNED,
    targetType: Entities.USER._type,
  },
  USER_OWNS_CMDB: {
    _type: 'service_now_user_owns_cmdb_object',
    sourceType: Entities.USER._type,
    _class: RelationshipClass.OWNS,
    targetType: Entities.CMDB_OBJECT._type,
  },
  USER_MANAGES_CMDB: {
    _type: 'service_now_user_manages_cmdb_object',
    sourceType: Entities.USER._type,
    _class: RelationshipClass.MANAGES,
    targetType: Entities.CMDB_OBJECT._type,
  },
  GROUP_MANAGES_CMDB: {
    _type: 'service_now_group_manages_cmdb_object',
    sourceType: Entities.GROUP._type,
    _class: RelationshipClass.MANAGES,
    targetType: Entities.CMDB_OBJECT._type,
  },
};
