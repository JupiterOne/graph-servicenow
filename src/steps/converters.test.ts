import {
  createUserEntity,
  createGroupEntity,
  createGroupUserRelationship,
  createGroupGroupRelationship,
  createAccountEntity,
  createIncidentEntity,
  createIncidentAssigneeRelationship,
} from './converters';
import { Entities } from '../constants';
import {
  ServiceNowActiveEnum,
  ServiceNowIncident,
  ServiceNowUser,
  ServiceNowGroup,
} from '../types';

test('createAccountEntity', () => {
  const hostname = 'dev00000.service-now.com';

  const accountEntity = createAccountEntity(hostname);

  expect(accountEntity).toMatchGraphObjectSchema({
    _class: [Entities.ACCOUNT._class],
    schema: {},
  });
  expect(accountEntity).toMatchSnapshot();
});

test('createUserEntity', () => {
  const user_password = 'some-bogus-password';

  const user: ServiceNowUser = {
    sys_id: '5ccbba87db1310109d87a9954b9619db',
    name: 'J1 Administrator',
    user_name: 'j1-administrator',
    active: ServiceNowActiveEnum.TRUE,
    email: 'j1-admin@jupiterone.com',
    user_password: user_password,
    sys_created_on: '2020-09-21 17:57:17',
    sys_updated_on: '2020-09-23 22:07:54',
  };

  const userEntity = createUserEntity(user);

  expect(userEntity).toMatchGraphObjectSchema({
    _class: [Entities.USER._class],
    schema: {},
  });
  expect(userEntity).toMatchSnapshot();
  expect(JSON.stringify(userEntity).includes(user_password)).toBe(false);
  expect(JSON.stringify(userEntity).includes('user_password')).toBe(false);
});

test('createGroupEntity', () => {
  const group: ServiceNowGroup = {
    name: 'test name',
    parent: {
      link:
        'https://dev94579.service-now.com/api/now/table/sys_user_group/43f6027fdb1710109d87a9954b961939',
      value: '43f6027fdb1710109d87a9954b961939',
    },
    active: ServiceNowActiveEnum.TRUE,
    sys_updated_on: '2020-09-23 22:03:48',
    sys_id: '66a74a7fdb1710109d87a9954b961914',
    sys_created_on: '2020-09-23 22:03:48',
    email: 'j1-subgroup@jupiterone.com',
  };

  const groupEntity = createGroupEntity(group);

  expect(groupEntity).toMatchGraphObjectSchema({
    _class: [Entities.GROUP._class],
    schema: {},
  });
  expect(groupEntity).toMatchSnapshot();

  const groupGroupRelationship = createGroupGroupRelationship(
    groupEntity,
    group.parent,
  );
  expect(groupGroupRelationship).toMatchSnapshot();
});

test('createGroupUserRelationship', () => {
  const groupUser = {
    sys_id: '94878e3fdb1710109d87a9954b96195f',
    sys_updated_by: 'j1-administrator',
    sys_created_on: '2020-09-23 22:02:55',
    sys_mod_count: '0',
    sys_updated_on: '2020-09-23 22:02:55',
    sys_tags: '',
    user: {
      link:
        'https://dev94579.service-now.com/api/now/table/sys_user/5ccbba87db1310109d87a9954b9619db',
      value: '5ccbba87db1310109d87a9954b9619db',
    },
    sys_created_by: 'j1-administrator',
    group: {
      link:
        'https://dev94579.service-now.com/api/now/table/sys_user_group/43f6027fdb1710109d87a9954b961939',
      value: '43f6027fdb1710109d87a9954b961939',
    },
  };

  const userGroupRelationship = createGroupUserRelationship(groupUser);

  expect(userGroupRelationship).toMatchSnapshot();
});

describe('createIncidentEntity', () => {
  const incident: ServiceNowIncident = {
    number: 'INC0000060',
    severity: '3',
    category: 'inquiry',
    opened_by: {
      link:
        'https://dev94579.service-now.com/api/now/table/sys_user/681ccaf9c0a8016400b98a06818d57c7',
      value: '681ccaf9c0a8016400b98a06818d57c7',
    },
    impact: 2,
    resolved_at: '2016-12-13 21:43:14',
    active: ServiceNowActiveEnum.FALSE,
    assigned_to: {
      link:
        'https://dev94579.service-now.com/api/now/table/sys_user/5137153cc611227c000bbd1bd8cd2007',
      value: '5137153cc611227c000bbd1bd8cd2007',
    },
    sys_id: '1c741bd70b2322007518478d83673af3',
    sys_created_on: '2016-12-12 15:19:57',
    sys_updated_on: '2016-12-14 02:46:44',
  };

  test('createIncidentEntity', () => {
    const incidentEntity = createIncidentEntity(incident);
    expect(incidentEntity).toMatchGraphObjectSchema({
      _class: [Entities.INCIDENT._class],
    });
    expect(incidentEntity).toMatchSnapshot();
  });

  test('createIncidentAssigneeRelationship', () => {
    const incidentAssigneeRelationship = createIncidentAssigneeRelationship(
      incident,
    );
    expect(incidentAssigneeRelationship).toMatchDirectRelationshipSchema({});
    expect(incidentAssigneeRelationship).toMatchSnapshot();
  });
});
