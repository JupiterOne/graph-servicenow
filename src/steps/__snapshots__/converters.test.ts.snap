// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAccountEntity 1`] = `
Object {
  "_class": Array [
    "Account",
  ],
  "_key": "dev00000.service-now.com",
  "_rawData": Array [],
  "_type": "service_now_account",
  "createdOn": undefined,
  "displayName": "dev00000.service-now.com",
  "name": "dev00000.service-now.com",
}
`;

exports[`createGroupEntity 1`] = `
Object {
  "_class": Array [
    "UserGroup",
  ],
  "_key": "service_now_group:66a74a7fdb1710109d87a9954b961914:j1-subgroup@jupiterone.com",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "active": "true",
        "cost_center": "",
        "default_assignee": "",
        "description": "",
        "email": "j1-subgroup@jupiterone.com",
        "exclude_manager": "false",
        "include_members": "false",
        "manager": "",
        "name": "j1-subgroup",
        "parent": Object {
          "link": "https://dev94579.service-now.com/api/now/table/sys_user_group/43f6027fdb1710109d87a9954b961939",
          "value": "43f6027fdb1710109d87a9954b961939",
        },
        "roles": "",
        "source": "",
        "sys_created_by": "j1-administrator",
        "sys_created_on": "2020-09-23 22:03:48",
        "sys_id": "66a74a7fdb1710109d87a9954b961914",
        "sys_mod_count": "0",
        "sys_tags": "",
        "sys_updated_by": "j1-administrator",
        "sys_updated_on": "2020-09-23 22:03:48",
        "type": "",
      },
    },
  ],
  "_type": "service_now_group",
  "active": true,
  "createdOn": 1600898628000,
  "description": "",
  "displayName": "j1-subgroup",
  "email": "j1-subgroup@jupiterone.com",
  "id": "66a74a7fdb1710109d87a9954b961914",
  "name": "j1-subgroup",
  "updatedOn": 1600898628000,
}
`;

exports[`createGroupEntity 2`] = `
Object {
  "_class": "HAS",
  "_fromEntityKey": "43f6027fdb1710109d87a9954b961939",
  "_key": "43f6027fdb1710109d87a9954b961939|has|service_now_group:66a74a7fdb1710109d87a9954b961914:j1-subgroup@jupiterone.com",
  "_toEntityKey": "service_now_group:66a74a7fdb1710109d87a9954b961914:j1-subgroup@jupiterone.com",
  "_type": "service_now_group_has_group",
  "displayName": "HAS",
}
`;

exports[`createGroupUserRelationship 1`] = `
Object {
  "_class": "HAS",
  "_fromEntityKey": "43f6027fdb1710109d87a9954b961939",
  "_key": "43f6027fdb1710109d87a9954b961939|has|5ccbba87db1310109d87a9954b9619db",
  "_toEntityKey": "5ccbba87db1310109d87a9954b9619db",
  "_type": "service_now_group_has_user",
  "createdOn": 1600898575000,
  "displayName": "HAS",
  "id": "94878e3fdb1710109d87a9954b96195f",
  "updatedOn": 1600898575000,
}
`;

exports[`createUserEntity 1`] = `
Object {
  "_class": Array [
    "User",
  ],
  "_key": "service_now_user:5ccbba87db1310109d87a9954b9619db:j1-administrator",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "active": "true",
        "avatar": "",
        "building": "",
        "calendar_integration": "1",
        "city": "",
        "company": "",
        "cost_center": "",
        "country": "",
        "date_format": "",
        "default_perspective": "",
        "department": Object {
          "link": "https://dev94579.service-now.com/api/now/table/cmn_department/9d7802bfdb1710109d87a9954b96199d",
          "value": "9d7802bfdb1710109d87a9954b96199d",
        },
        "email": "j1-admin@jupiterone.com",
        "employee_number": "",
        "enable_multifactor_authn": "false",
        "failed_attempts": "0",
        "first_name": "Jupiterone",
        "gender": "",
        "home_phone": "",
        "internal_integration_user": "false",
        "introduction": "",
        "last_login": "2020-09-23",
        "last_login_time": "2020-09-23 22:00:16",
        "last_name": "Administrator",
        "ldap_server": "",
        "location": "",
        "locked_out": "false",
        "manager": "",
        "middle_name": "",
        "mobile_phone": "",
        "name": "Jupiterone Administrator",
        "notification": "2",
        "password_needs_reset": "false",
        "phone": "",
        "photo": "",
        "preferred_language": "",
        "roles": "",
        "schedule": "",
        "source": "",
        "state": "",
        "street": "",
        "sys_class_name": "sys_user",
        "sys_created_by": "admin",
        "sys_created_on": "2020-09-21 17:57:17",
        "sys_domain": Object {
          "link": "https://dev94579.service-now.com/api/now/table/sys_user_group/global",
          "value": "global",
        },
        "sys_domain_path": "/",
        "sys_id": "5ccbba87db1310109d87a9954b9619db",
        "sys_mod_count": "5",
        "sys_tags": "",
        "sys_updated_by": "j1-administrator",
        "sys_updated_on": "2020-09-23 22:07:54",
        "time_format": "",
        "time_zone": "",
        "title": "",
        "user_name": "j1-administrator",
        "vip": "false",
        "web_service_access_only": "false",
        "zip": "",
      },
    },
  ],
  "_type": "service_now_user",
  "active": true,
  "createdOn": 1600711037000,
  "displayName": "Jupiterone Administrator",
  "email": "j1-admin@jupiterone.com",
  "id": "5ccbba87db1310109d87a9954b9619db",
  "name": "Jupiterone Administrator",
  "updatedOn": 1600898874000,
  "username": "j1-administrator",
}
`;

exports[`incident createIncidentAssigneeRelationship 1`] = `
Object {
  "_class": "ASSIGNED",
  "_fromEntityKey": "1c741bd70b2322007518478d83673af3",
  "_key": "1c741bd70b2322007518478d83673af3|assigned|5137153cc611227c000bbd1bd8cd2007",
  "_toEntityKey": "5137153cc611227c000bbd1bd8cd2007",
  "_type": "service_now_incident_assigned_user",
  "displayName": "ASSIGNED",
}
`;

exports[`incident createIncidentEntity 1`] = `
Object {
  "_class": Array [
    "Incident",
  ],
  "_key": "service_now_incident:1c741bd70b2322007518478d83673af3:INC0000060",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "active": "false",
        "activity_due": "2016-12-13 01:26:36",
        "additional_assignee_list": "",
        "approval": "not requested",
        "approval_history": "",
        "approval_set": "",
        "assigned_to": Object {
          "link": "https://dev94579.service-now.com/api/now/table/sys_user/5137153cc611227c000bbd1bd8cd2007",
          "value": "5137153cc611227c000bbd1bd8cd2007",
        },
        "assignment_group": Object {
          "link": "https://dev94579.service-now.com/api/now/table/sys_user_group/287ebd7da9fe198100f92cc8d1d2154e",
          "value": "287ebd7da9fe198100f92cc8d1d2154e",
        },
        "business_duration": "1970-01-01 08:00:00",
        "business_service": Object {
          "link": "https://dev94579.service-now.com/api/now/table/cmdb_ci_service/27d32778c0a8000b00db970eeaa60f16",
          "value": "27d32778c0a8000b00db970eeaa60f16",
        },
        "business_stc": "28800",
        "calendar_duration": "1970-01-02 04:23:17",
        "calendar_stc": "102197",
        "caller_id": Object {
          "link": "https://dev94579.service-now.com/api/now/table/sys_user/681ccaf9c0a8016400b98a06818d57c7",
          "value": "681ccaf9c0a8016400b98a06818d57c7",
        },
        "category": "inquiry",
        "caused_by": "",
        "child_incidents": "0",
        "close_code": "Solved (Permanently)",
        "close_notes": "This incident is resolved.",
        "closed_at": "2016-12-14 02:46:44",
        "closed_by": Object {
          "link": "https://dev94579.service-now.com/api/now/table/sys_user/681ccaf9c0a8016400b98a06818d57c7",
          "value": "681ccaf9c0a8016400b98a06818d57c7",
        },
        "cmdb_ci": Object {
          "link": "https://dev94579.service-now.com/api/now/table/cmdb_ci/109562a3c611227500a7b7ff98cc0dc7",
          "value": "109562a3c611227500a7b7ff98cc0dc7",
        },
        "comments": "",
        "comments_and_work_notes": "",
        "company": Object {
          "link": "https://dev94579.service-now.com/api/now/table/core_company/31bea3d53790200044e0bfc8bcbe5dec",
          "value": "31bea3d53790200044e0bfc8bcbe5dec",
        },
        "contact_type": "self-service",
        "contract": "",
        "correlation_display": "",
        "correlation_id": "",
        "delivery_plan": "",
        "delivery_task": "",
        "description": "I am unable to connect to the email server. It appears to be down.",
        "due_date": "",
        "escalation": "0",
        "expected_start": "",
        "follow_up": "",
        "group_list": "",
        "hold_reason": "",
        "impact": "2",
        "incident_state": "7",
        "knowledge": "false",
        "location": "",
        "made_sla": "true",
        "notify": "1",
        "number": "INC0000060",
        "opened_at": "2016-12-12 15:19:57",
        "opened_by": Object {
          "link": "https://dev94579.service-now.com/api/now/table/sys_user/681ccaf9c0a8016400b98a06818d57c7",
          "value": "681ccaf9c0a8016400b98a06818d57c7",
        },
        "order": "",
        "parent": "",
        "parent_incident": "",
        "priority": "3",
        "problem_id": "",
        "reassignment_count": "2",
        "reopen_count": "0",
        "reopened_by": "",
        "reopened_time": "",
        "resolved_at": "2016-12-13 21:43:14",
        "resolved_by": Object {
          "link": "https://dev94579.service-now.com/api/now/table/sys_user/5137153cc611227c000bbd1bd8cd2007",
          "value": "5137153cc611227c000bbd1bd8cd2007",
        },
        "rfc": "",
        "route_reason": "",
        "service_offering": "",
        "severity": "3",
        "short_description": "Unable to connect to email",
        "sla_due": "",
        "state": "7",
        "subcategory": "email",
        "sys_class_name": "incident",
        "sys_created_by": "employee",
        "sys_created_on": "2016-12-12 15:19:57",
        "sys_domain": Object {
          "link": "https://dev94579.service-now.com/api/now/table/sys_user_group/global",
          "value": "global",
        },
        "sys_domain_path": "/",
        "sys_id": "1c741bd70b2322007518478d83673af3",
        "sys_mod_count": "15",
        "sys_tags": "",
        "sys_updated_by": "employee",
        "sys_updated_on": "2016-12-14 02:46:44",
        "task_effective_number": "INC0000060",
        "time_worked": "",
        "universal_request": "",
        "upon_approval": "proceed",
        "upon_reject": "cancel",
        "urgency": "2",
        "user_input": "",
        "watch_list": "",
        "work_end": "",
        "work_notes": "",
        "work_notes_list": "",
        "work_start": "",
      },
    },
  ],
  "_type": "service_now_incident",
  "active": false,
  "category": "inquiry",
  "createdOn": 1481555997000,
  "description": "I am unable to connect to the email server. It appears to be down.",
  "displayName": "INC0000060",
  "id": "1c741bd70b2322007518478d83673af3",
  "impact": "2",
  "name": "INC0000060",
  "reportable": false,
  "reporter": "681ccaf9c0a8016400b98a06818d57c7",
  "resolvedAt": "2016-12-13 21:43:14",
  "severity": "3",
  "updatedOn": 1481683604000,
}
`;
